sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -4034,6 +4034,11 @@ pi_result piEnqueueEventsWaitWithBarrier(pi_queue Queue,
                                      (*Event)->WaitList.Length,
                                      (*Event)->WaitList.ZeEventList));
 
+  // Execute command list asynchronously as the event will be used
+  // to track down its completion.
+  if (auto Res = Queue->executeCommandList(ZeCommandList, ZeFence))
+    return Res;
+
```

Comment:
Have you taken a look at the failed test?
I am not sure if we can execute the command list without checking whether it has been executed already.
For example, executing a = a+1 will generate an incorrect result.

By the way, you can simply return the result of Queue->executeCommandList() instead of checking if the result is success or not, i.e.,  `return Queue->executeCommandList(ZeCommandList, ZeFence);`

Reply:
> Have you taken a look at the failed test?

Yes, I updated the patch to force submission of the command lists that this barrier is dependent on.

> I am not sure if we can execute the command list without checking whether it has been executed already. 

Without this patch the barrier was never sent to gpu. It is not executing twice.

Reply:
Yes, I just meant that the dependent commands in the waitlist should not be executed twice, not the barrier itself.
Your latest patch fixed my concern.